wtxmgr/db.go

```diff
@@ -0,0 +1,1311 @@
+			return storeError(ErrDatabase, str, err)
+		}
+
+		_, err = ns.CreateBucketIfNotExists(bucketCredits)
+		if err != nil {
+			str := "failed to create credits bucket"
+			return storeError(ErrDatabase, str, err)
+		}
+
+		_, err = ns.CreateBucketIfNotExists(bucketDebits)
+		if err != nil {
+			str := "failed to create debits bucket"
+			return storeError(ErrDatabase, str, err)
+		}
+
+		_, err = ns.CreateBucketIfNotExists(bucketUnspent)
+		if err != nil {
+			str := "failed to create unspent bucket"
+			return storeError(ErrDatabase, str, err)
+		}
+
+		_, err = ns.CreateBucketIfNotExists(bucketUnmined)
+		if err != nil {
+			str := "failed to create unmined bucket"
+			return storeError(ErrDatabase, str, err)
+		}
+
+		_, err = ns.CreateBucketIfNotExists(bucketUnminedCredits)
+		if err != nil {
+			str := "failed to create unmined credits bucket"
+			return storeError(ErrDatabase, str, err)
+		}
+
+		_, err = ns.CreateBucketIfNotExists(bucketUnminedInputs)
+		if err != nil {
+			str := "failed to create unmined inputs bucket"
+			return storeError(ErrDatabase, str, err)
+		}
+
+		// Save the most recent tx store version if it isn't already
+		// there, otherwise keep track of it for potential upgrades.
+		v := ns.Get(rootVersion)
+		if len(v) != 4 {
+			v = make([]byte, 4)
+			byteOrder.PutUint32(v, version)
+			err := ns.Put(rootVersion, v)
+			if err != nil {
+				str := "failed to store latest database version"
+				return storeError(ErrDatabase, str, err)
+			}
+			version = LatestTxStoreVersion
+		} else {
+			version = byteOrder.Uint32(v)
+		}
+
+		// Save the creation date of the store if it does not already
+		// exist.
+		v = ns.Get(rootCreateDate)
+		if v == nil {
+			v = make([]byte, 8)
+			byteOrder.PutUint64(v, uint64(time.Now().Unix()))
+			err := ns.Put(rootCreateDate, v)
+			if err != nil {
+				str := "failed to store database creation time"
+				return storeError(ErrDatabase, str, err)
+			}
+		}
+
+		// Write a zero balance if it does not already exist.
+		v = ns.Get(rootMinedBalance)
+		if v == nil {
+			v = make([]byte, 8)
+			err := ns.Put(rootMinedBalance, v)
+			if err != nil {
+				str := "failed to write zero balance"
+				return storeError(ErrDatabase, str, err)
+			}
+		}
+
+		return nil
+	})
+	if err != nil {
+		str := "failed to update database"
+		return storeError(ErrDatabase, str, err)
+	}
+
+	// Upgrade the tx store as needed.
+	// No upgrades yet.
+	//if version < LatestTxStoreVersion {
+	//}
+
+	return nil
+}
+
+func scopedUpdate(ns walletdb.Namespace, f func(walletdb.Bucket) error) error {
+	tx, err := ns.Begin(true)
+	if err != nil {
+		str := "cannot begin update"
+		return storeError(ErrDatabase, str, err)
+	}
+	err = f(tx.RootBucket())
+	if err != nil {
+		rbErr := tx.Rollback()
+		if rbErr != nil {
+			str := "rollback failed"
+			return storeError(ErrDatabase, str, rbErr)
+		}
+		return err
+	}
+	err = tx.Commit()
+	if err != nil {
+		str := "commit failed"
```

Comment:
We need to check how bolt handles a commit failure.  I think it might require a call to `.Rollback` in this scenario to release the page references associated with it, but I'm not 100% sure on that.